File: lib/pro/workbench-panel.js
function asArray(value = []) {
  return Array.isArray(value) ? value.filter(Boolean) : [];
}

function compactString(value) {
  const text = String(value ?? "").trim();
  return text || null;
}

function clampProgress(value) {
  if (typeof value !== "number" || Number.isNaN(value)) return null;
  return Math.max(0, Math.min(100, Math.round(value)));
}

function normalizeAction(action = {}) {
  const id = compactString(action.id || action.intent || action.label);
  if (!id) return null;
  return {
    id,
    label: compactString(action.label || action.intent || action.id) || "Action",
    intent: compactString(action.intent),
    type: compactString(action.type) || "action",
    command: compactString(action.command),
    value: compactString(action.value),
    enabled: action.enabled !== false,
    placement: compactString(action.placement) || "secondary",
    requiresUserApproval: Boolean(action.requiresUserApproval)
  };
}

function normalizeRows(rows = []) {
  return asArray(rows)
    .map(row => ({
      label: compactString(row.label || row.key || row.id) || "Value",
      value: compactString(row.value ?? row.text ?? row.summary) || ""
    }))
    .filter(row => row.label || row.value);
}

function normalizeSectionItem(item = {}) {
  return {
    title: compactString(item.title || item.code || item.id) || "Item",
    code: compactString(item.code),
    status: compactString(item.status || item.state),
    command: compactString(item.command),
    surface: compactString(item.surface),
    expectedEvidence: compactString(item.expectedEvidence || item.summary || item.value)
  };
}

function normalizeSection(section = {}) {
  const id = compactString(section.id || section.type || section.title);
  if (!id) return null;
  const actions = asArray(section.actions).map(normalizeAction).filter(Boolean);
  return {
    id,
    title: compactString(section.title || section.id) || "Section",
    type: compactString(section.type) || "section",
    status: compactString(section.status || section.state),
    summary: compactString(section.summary || section.description),
    target: compactString(section.target),
    command: compactString(section.command),
    rows: normalizeRows(section.rows),
    items: asArray(section.items).map(normalizeSectionItem).filter(Boolean),
    actions
  };
}

function normalizeNavCard(card = {}) {
  const id = compactString(card.id || card.title);
  if (!id) return null;
  return {
    id,
    label: compactString(card.title || card.id) || "Card",
    value: compactString(card.value) || "",
    state: compactString(card.state),
    description: compactString(card.description || card.summary),
    progress: clampProgress(card.progress)
  };
}

function findAction(actions = [], predicate) {
  return asArray(actions).find(predicate) || null;
}

function firstSentence(value) {
  const text = compactString(value);
  if (!text) return null;
  const sentence = text.split(/(?<=[.!?])\s+/)[0] || text;
  return sentence.length > 180 ? `${sentence.slice(0, 177)}...` : sentence;
}

function joinParts(parts = [], separator = " | ") {
  return parts.map(compactString).filter(Boolean).join(separator);
}

function renderStatusStrip(cards = []) {
  const values = asArray(cards).slice(0, 4).map(card => {
    const state = card.state ? `/${card.state}` : "";
    const progress = typeof card.progress === "number" ? ` ${card.progress}%` : "";
    return `${card.label}: ${card.value}${state}${progress}`.trim();
  });
  return values.length > 0 ? values.join("  |  ") : "No status cards yet";
}

function renderSectionSummary(section = {}) {
  const pieces = [
    section.status ? `[${section.status}]` : null,
    section.title,
    firstSentence(section.summary)
  ];
  return joinParts(pieces, " | ");
}

export function buildWorkbenchPanelModel(view = null, { locale = null } = {}) {
  if (!view || typeof view !== "object") return null;

  const decision = view.decision || {};
  const primaryActions = asArray(view.primaryActions).map(normalizeAction).filter(Boolean);
  const sections = asArray(view.sections).map(normalizeSection).filter(Boolean);
  const navCards = asArray(view.navigation).map(normalizeNavCard).filter(Boolean);
  const agentLoop = view.agentLoop || {};
  const nextLoopPackage = agentLoop.nextLoopPackage || {};
  const nextProofSection = sections.find(section => section.id === "next-proof") || null;
  const executeAction = findAction(primaryActions, action =>
    action.intent === "execute_next_proof" || action.intent === "execute_proof_command" || Boolean(action.command)
  ) || findAction(nextProofSection?.actions, action => Boolean(action.command));
  const handoffAction = findAction(primaryActions, action => action.intent === "handoff_to_agent");

  return {
    format: "agent-guardrails-workbench-panel.v1",
    schemaVersion: 1,
    sourceFormat: compactString(view.format) || "agent-guardrails-workbench-view.v1",
    locale: compactString(locale) || "en",
    renderer: "native-panel-model",
    generatedFor: ["desktop-app", "native-panel", "mcp-client", "terminal-ui"],
    displayRules: {
      primarySurface: compactString(view.renderHints?.primarySurface) || "native-panel",
      density: compactString(view.renderHints?.density) || "compact",
      tone: compactString(view.renderHints?.tone || decision.state) || "unknown",
      avoid: asArray(view.renderHints?.avoid).map(String),
      rawJsonIsSupportingDetail: true,
      htmlIsLegacyFallback: true
    },
    hero: {
      question: compactString(decision.question || view.headline) || "Can I ship this change?",
      answer: compactString(decision.answer || decision.verdict) || "Unknown",
      title: compactString(decision.title) || "Decision",
      state: compactString(decision.state || decision.verdict) || "unknown",
      riskLabel: compactString(decision.riskLabel || decision.riskTier),
      mergeGate: compactString(decision.mergeGate),
      trustScore: typeof decision.trustScore === "number" ? decision.trustScore : null,
      reason: compactString(decision.reason || view.summary)
    },
    statusStrip: navCards,
    nextStep: {
      label: compactString(executeAction?.label || nextProofSection?.title) || "Next step",
      command: compactString(nextLoopPackage.nextCommand || executeAction?.command || nextProofSection?.command),
      intent: compactString(executeAction?.intent) || "execute_next_proof",
      summary: compactString(executeAction?.value || nextProofSection?.summary),
      rerunCommand: compactString(nextLoopPackage.rerunCommand || agentLoop.rerunCommand),
      evidenceTool: compactString(agentLoop.evidenceCapture?.tool || nextLoopPackage.evidenceCapture?.tool)
    },
    handoff: {
      label: compactString(handoffAction?.label) || "Copy agent handoff",
      intent: compactString(handoffAction?.intent) || "handoff_to_agent",
      brief: compactString(handoffAction?.value || nextLoopPackage.agentBrief),
      humanRole: compactString(agentLoop.humanRole),
      humanChecks: asArray(agentLoop.humanChecks).map(String),
      stopConditions: asArray(agentLoop.stopConditions).map(String)
    },
    actions: primaryActions,
    sections,
    motion: {
      principle: compactString(view.renderHints?.motion?.principle) || "Use calm state transitions; do not animate raw logs.",
      transitions: asArray(view.renderHints?.motion?.transitions).map(String)
    },
    emptyStates: asArray(view.renderHints?.emptyStates),
    legacy: {
      htmlRenderer: compactString(view.legacy?.htmlRenderer),
      status: compactString(view.legacy?.status) || "compatibility_only"
    }
  };
}

export function renderWorkbenchPanelText(panel = null) {
  if (!panel || typeof panel !== "object") {
    return "Agent Guardrails Workbench\nNo panel model is available yet.";
  }

  const hero = panel.hero || {};
  const nextStep = panel.nextStep || {};
  const handoff = panel.handoff || {};
  const sections = asArray(panel.sections).slice(0, 4);
  const header = joinParts([
    "Agent Guardrails Workbench",
    hero.state ? hero.state.toUpperCase() : null,
    hero.riskLabel
  ]);
  const lines = [
    header,
    "=".repeat(Math.min(72, Math.max(32, header.length))),
    `${hero.question || "Can I ship this change?"} ${hero.answer ? `-> ${hero.answer}` : ""}`.trim(),
    hero.reason ? `Reason: ${firstSentence(hero.reason)}` : null,
    typeof hero.trustScore === "number" ? `Trust score: ${hero.trustScore}` : null,
    panel.statusStrip?.length ? `Status: ${renderStatusStrip(panel.statusStrip)}` : null,
    "",
    "Next step",
    "-".repeat(9),
    nextStep.label ? `Action: ${nextStep.label}` : null,
    nextStep.summary ? `Why: ${firstSentence(nextStep.summary)}` : null,
    nextStep.command ? `$ ${nextStep.command}` : null,
    nextStep.rerunCommand ? `Then rerun: ${nextStep.rerunCommand}` : null,
    nextStep.evidenceTool ? `Evidence tool: ${nextStep.evidenceTool}` : null,
    "",
    "Agent handoff",
    "-".repeat(13),
    handoff.label ? `Action: ${handoff.label}` : null,
    handoff.humanRole ? `Human role: ${firstSentence(handoff.humanRole)}` : null,
    handoff.humanChecks?.length ? `Check: ${handoff.humanChecks.slice(0, 2).join(" | ")}` : null,
    handoff.stopConditions?.length ? `Stop if: ${handoff.stopConditions.slice(0, 2).join(" | ")}` : null,
    "",
    "Sections",
    "-".repeat(8),
    ...sections.map(section => `- ${renderSectionSummary(section)}`)
  ];

  return lines.filter(line => line !== null).join("\n").trimEnd();
}

File: crates/agent-guardrails-cli/src/workbench_panel.rs
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

const DEFAULT_PANEL_PATH: &str = ".agent-guardrails/pro/operator-workbench-panel.json";

pub fn run_workbench_panel_cli(args: &[String]) -> i32 {
    let options = match parse_args(args) {
        Ok(value) => value,
        Err(error) => {
            eprintln!("agent-guardrails workbench-panel: {error}");
            return 1;
        }
    };

    let panel = match read_panel(&options.file) {
        Ok(value) => value,
        Err(error) => {
            eprintln!("agent-guardrails workbench-panel: {error}");
            return 1;
        }
    };

    if options.json {
        match serde_json::to_string_pretty(&panel) {
            Ok(json) => println!("{json}"),
            Err(error) => {
                eprintln!(
                    "agent-guardrails workbench-panel: failed to serialize panel JSON: {error}"
                );
                return 1;
            }
        }
        return 0;
    }

    println!("{}", render_workbench_panel_text(&panel));
    0
}

struct PanelOptions {
    file: PathBuf,
    json: bool,
}

fn parse_args(args: &[String]) -> Result<PanelOptions, String> {
    let mut file = PathBuf::from(DEFAULT_PANEL_PATH);
    let mut json = false;
    let mut index = 0;

    while index < args.len() {
        match args[index].as_str() {
            "--file" | "--from" => {
                let value = args
                    .get(index + 1)
                    .filter(|item| !item.starts_with('-'))
                    .ok_or_else(|| format!("missing value for {}", args[index]))?;
                file = PathBuf::from(value);
                index += 2;
            }
            "--json" => {
                json = true;
                index += 1;
            }
            "--help" | "-h" => {
                return Err(
                    "usage: agent-guardrails workbench-panel [--file <operator-workbench-panel.json>] [--json]"
                        .to_string(),
                );
            }
            value if value.starts_with('-') => return Err(format!("unknown option: {value}")),
            value => return Err(format!("unexpected argument: {value}")),
        }
    }

    Ok(PanelOptions { file, json })
}

fn read_panel(file: &PathBuf) -> Result<Value, String> {
    let raw = fs::read_to_string(file)
        .map_err(|error| format!("failed to read {}: {error}", file.display()))?;
    let value: Value = serde_json::from_str(raw.trim_start_matches('\u{feff}'))
        .map_err(|error| format!("failed to parse {} as JSON: {error}", file.display()))?;
    let format = value
        .get("format")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if format != "agent-guardrails-workbench-panel.v1" {
        return Err(format!(
            "unsupported panel format {format:?}; expected agent-guardrails-workbench-panel.v1"
        ));
    }
    Ok(value)
}

pub fn render_workbench_panel_text(panel: &Value) -> String {
    let header_model = panel.get("header").unwrap_or(&Value::Null);
    let trust = panel.get("trust").unwrap_or(&Value::Null);
    let hero = panel
        .get("hero")
        .or_else(|| panel.get("decision"))
        .unwrap_or(&Value::Null);
    let next_step = panel.get("nextStep").unwrap_or(&Value::Null);
    let handoff = panel
        .get("handoff")
        .or_else(|| panel.get("agentHandoff"))
        .unwrap_or(&Value::Null);
    let header = join_parts(&[
        string_field(header_model, "title")
            .or_else(|| string_field(hero, "title"))
            .or_else(|| Some("Agent Guardrails Workbench".to_string())),
        string_field(hero, "state")
            .or_else(|| string_field(header_model, "status"))
            .map(|value| value.to_uppercase()),
        string_field(hero, "riskLabel").or_else(|| string_field(header_model, "severity")),
    ]);
    let mut lines = vec![
        header.clone(),
        "=".repeat(header.len().clamp(32, 72)),
        format!(
            "{}{}",
            string_field(hero, "question").unwrap_or_else(|| "Can I ship this change?".to_string()),
            string_field(hero, "answer")
                .map(|answer| format!(" -> {answer}"))
                .unwrap_or_default()
        ),
    ];

    if let Some(reason) = string_field(hero, "reason").and_then(|value| first_sentence(&value)) {
        lines.push(format!("Reason: {reason}"));
    }
    if let Some(score) = hero
        .get("trustScore")
        .and_then(Value::as_i64)
        .or_else(|| trust.get("score").and_then(Value::as_i64))
    {
        lines.push(format!("Trust score: {score}"));
    }
    if let Some(status) = render_status_strip(panel.get("statusStrip").and_then(Value::as_array)) {
        lines.push(format!("Status: {status}"));
    }

    lines.extend([
        "".to_string(),
        "Next step".to_string(),
        "---------".to_string(),
    ]);
    push_labeled(
        &mut lines,
        "Action",
        string_field(next_step, "label").or_else(|| string_field(next_step, "title")),
    );
    push_labeled(
        &mut lines,
        "Why",
        string_field(next_step, "summary")
            .or_else(|| string_field(next_step, "body"))
            .and_then(|value| first_sentence(&value)),
    );
    if let Some(command) = string_field(next_step, "command") {
        lines.push(format!("$ {command}"));
    }
    push_labeled(
        &mut lines,
        "Then rerun",
        string_field(next_step, "rerunCommand"),
    );
    push_labeled(
        &mut lines,
        "Evidence tool",
        string_field(next_step, "evidenceTool"),
    );

    lines.extend([
        "".to_string(),
        "Agent handoff".to_string(),
        "-------------".to_string(),
    ]);
    push_labeled(
        &mut lines,
        "Action",
        string_field(handoff, "label")
            .or_else(|| string_field(handoff, "title"))
            .or_else(|| string_field(handoff, "copyLabel")),
    );
    push_labeled(
        &mut lines,
        "Human role",
        string_field(handoff, "humanRole")
            .or_else(|| string_field(handoff, "summary"))
            .and_then(|value| first_sentence(&value)),
    );
    if let Some(checks) =
        render_string_array(handoff.get("humanChecks").and_then(Value::as_array), 2)
    {
        lines.push(format!("Check: {checks}"));
    }
    if let Some(stop) =
        render_string_array(handoff.get("stopConditions").and_then(Value::as_array), 2)
    {
        lines.push(format!("Stop if: {stop}"));
    }

    lines.extend([
        "".to_string(),
        "Sections".to_string(),
        "--------".to_string(),
    ]);
    for section in panel
        .get("sections")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .take(4)
    {
        let parts = join_parts(&[
            string_field(section, "status")
                .or_else(|| string_field(section, "state"))
                .map(|status| format!("[{status}]")),
            string_field(section, "title"),
            string_field(section, "summary").and_then(|value| first_sentence(&value)),
        ]);
        if !parts.is_empty() {
            lines.push(format!("- {parts}"));
        }
    }

    lines
        .into_iter()
        .filter(|line| !line.trim().is_empty() || line.is_empty())
        .collect::<Vec<_>>()
        .join("\n")
}

fn push_labeled(lines: &mut Vec<String>, label: &str, value: Option<String>) {
    if let Some(value) = value {
        lines.push(format!("{label}: {value}"));
    }
}

fn string_field(value: &Value, key: &str) -> Option<String> {
    value
        .get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|item| !item.is_empty())
        .map(ToString::to_string)
}

fn first_sentence(value: &str) -> Option<String> {
    let text = value.trim();
    if text.is_empty() {
        return None;
    }
    let mut split_at = text.len();
    for marker in ['.', '!', '?'] {
        if let Some(index) = text.find(marker) {
            split_at = split_at.min(index + marker.len_utf8());
        }
    }
    let sentence = text[..split_at].trim();
    let char_count = sentence.chars().count();
    if char_count > 180 {
        Some(format!(
            "{}...",
            sentence.chars().take(177).collect::<String>()
        ))
    } else {
        Some(sentence.to_string())
    }
}

fn join_parts(parts: &[Option<String>]) -> String {
    parts
        .iter()
        .filter_map(|part| part.as_deref())
        .filter(|part| !part.trim().is_empty())
        .collect::<Vec<_>>()
        .join(" | ")
}

fn render_string_array(values: Option<&Vec<Value>>, limit: usize) -> Option<String> {
    let text = values?
        .iter()
        .filter_map(Value::as_str)
        .map(str::trim)
        .filter(|item| !item.is_empty())
        .take(limit)
        .collect::<Vec<_>>()
        .join(" | ");
    if text.is_empty() { None } else { Some(text) }
}

fn render_status_strip(values: Option<&Vec<Value>>) -> Option<String> {
    let rendered = values?
        .iter()
        .take(4)
        .filter_map(|card| {
            let label = string_field(card, "label")?;
            let value = string_field(card, "value").unwrap_or_default();
            let state = string_field(card, "state")
                .map(|state| format!("/{state}"))
                .unwrap_or_default();
            let progress = card
                .get("progress")
                .and_then(Value::as_i64)
                .map(|value| format!(" {value}%"))
                .unwrap_or_default();
            Some(
                format!("{label}: {value}{state}{progress}")
                    .trim()
                    .to_string(),
            )
        })
        .collect::<Vec<_>>()
        .join("  |  ");
    if rendered.is_empty() {
        None
    } else {
        Some(rendered)
    }
}
